### PR TITLE
Feature/inserm/narrative

### PR DESCRIPTION
**Pull request by gregoire-pointeau**
Input from INSERM concerning narrative-related work for WYSIWYD Y3.

### CODE REVIEW

**Comment by Tobias-Fischer** [comment]
Hi @gregoire-pointeau,
Thanks for the PR! It's quite big - I won't look into too much detail, especially as everything is only related to the narrative part and doesn't change anything else.
Two comments:
1) Please remove tabs with spaces ........
2) Please resolve the conflict in `opcPopulator.cpp` so we can merge
3) It would be great if you could replace some important `cout` with `yDebug`/`yInfo`/`yWarning` etc.

Thanks :)

**Comment by Tobias-Fischer** [comment]
Hi @gregoire-pointeau, did you manage to resolve the conflict? Do you need help?

Best, Tobi

**Comment by gregoire-pointeau** [comment]
It's okay, I'm on it right now. I'll tell you once it's done. Thanks !

**Comment by gregoire-pointeau** [comment]
I guess everything is good now.
git rebase --continue gives: `No rebase in progress`

Should I `git push` ?

**Comment by pattacini** [comment]
`git push --force-with-lease` 'cause you need to change the history on GitHub.
👍

**Comment by pattacini** [comment]
Forget! Just `git push` should be sufficient

**Comment by gregoire-pointeau** [comment]
Ok, I guess it's done.

**Comment by Tobias-Fischer** [comment]
Not quite :P

**Comment by gregoire-pointeau** [comment]
Hum, 

> $ git status
On branch devel
Your branch is up-to-date with 'origin/devel'.
nothing to commit, working directory clean

**Comment by Tobias-Fischer** [comment]
You should be in branch `feature/inserm/narrative` ..

**Comment by Tobias-Fischer** [comment]
Ah gosh. You pushed directly to `devel` .....

**Comment by Tobias-Fischer** [comment]
I have no idea whether this has broken anything. @pattacini would you please have a look?

**Comment by pattacini** [comment]
I've just reverted everything: `devel` is now safe as it was before.
@gregoire-pointeau you have to rebase your branch on top of devel, not the other way round 😄

**Comment by pattacini** [comment]
Let's do the big rebase together on Slack just after lunch.
I'd need your help to resolve conflicts.

**Comment by gregoire-pointeau** [comment]
Thanks to Ugo, you should be able to go Tobias.

**Comment by pattacini** [comment]
Hi @robotology/wysiwyd-developers 

We have this nice environment https://ide.c9.io/pattacini/wysiwyd to for example help people do tough rebasing in a collaborative way 😉 

It's just a matter of subscribing to [**C9**](https://c9.io) (it's free and requires only your GitHub account).

File: main/src/modules/abm/abmReasoning/src/abmReasoning.cpp [hunk 1]
```diff
@@ -186,6 +186,8 @@ bool abmReasoning::respond(const yarp::os::Bottle& bCommand, yarp::os::Bottle& b
 {
 
     bReply.clear();
+    bReply.addVocab(Vocab::encode("many"));
```
Comment (Tobias-Fischer) [inline main/src/modules/abm/abmReasoning/src/abmReasoning.cpp]: Hi @gregoire-pointeau,
What is this used for? Was this added by mistake? Does it break compatibility with other modules which interact with `abmReasoning`?

File: main/src/modules/abm/autobiographicalMemory/src/autobiographicalMemory.cpp [hunk 2]
```diff
@@ -446,6 +446,8 @@ bool autobiographicalMemory::respond(const Bottle& bCommand, Bottle& bReply)
 {
     Bottle bError;
     bReply.clear();
+    bReply.addVocab(Vocab::encode("many"));
```
Comment (Tobias-Fischer) [inline main/src/modules/abm/autobiographicalMemory/src/autobiographicalMemory.cpp]: Same as in `abmReasoning` - not sure what this is used for. However, I am quite concerned that the change here breaks quite a few other modules.

File: main/src/modules/attentionRelated/proactiveTagging/src/proactiveTagging.cpp [hunk 3]
```diff
@@ -172,6 +172,8 @@ bool proactiveTagging::respond(const Bottle& command, Bottle& reply) {
         "exploreKinematicByJoint joint bodypart [true/false] \n";
 
     reply.clear();
+    reply.addVocab(Vocab::encode("many"));
```
Comment (Tobias-Fischer) [inline main/src/modules/attentionRelated/proactiveTagging/src/proactiveTagging.cpp]: Same again.

File: main/src/modules/reservoir/reservoirHandler/src/main.cpp [hunk 4]
```diff
@@ -15,30 +15,21 @@
...
-    //test to launch a PDDL planner
-    //system("cd ~/Robots/planner/problem/ && ../seq-sat-lama-2011/plan efaa-dom.pddl efaa-prob.pddl res_plan.txt") ;
-    //system("cd ~/Robots/planner/problem/ && mv res_plan.txt* res_plan.txt") ;
+	yarp::os::Network yarp;
+	reservoirHandler module;
 
 
     /* prepare and configure the resource finder */
-    ResourceFinder rf;
+	yarp::os::ResourceFinder rf;
```
Comment (Tobias-Fischer) [inline main/src/modules/reservoir/reservoirHandler/src/main.cpp]: Tab -> space

File: main/src/modules/reservoir/reservoirHandler/include/reservoirHandler.h [hunk 5]
```diff
@@ -127,20 +117,20 @@ class reservoirHandler : public RFModule {
...
 
     int copyTrainData(const char* fileNameIn, const char* fileNameOut);
     int copyPastFile(const char* in, const char* fileNameOut);
     int trainSaveMeaningSentence(const char *filename);
-    int createTestwithTrainData(const char* filename, string sMeaningSentence);
-    string openResult(const char* fileNameIn);
-    bool AREactions(vector<string> seq);
-    vector<string> extractVocabulary(string sequence);
+	int createTestwithTrainData(const char* filename, std::string sMeaningSentence);
+	std::string openResult(const char* fileNameIn);
```
Comment (Tobias-Fischer) [inline main/src/modules/reservoir/reservoirHandler/include/reservoirHandler.h]: Tab -> space

File: main/src/modules/reservoir/reservoirHandler/src/reservoirHandler.cpp [hunk 6]
```diff
@@ -36,6 +33,7 @@ reservoirHandler::~reservoirHandler()
 
 bool reservoirHandler::configure(ResourceFinder &rf) {
 
+	iCurrentInstance = -1;
```
Comment (Tobias-Fischer) [inline main/src/modules/reservoir/reservoirHandler/src/reservoirHandler.cpp]: Tab -> space

File: main/src/modules/tools/InsermInteractionEngine/src/iie.cpp [hunk 7]
```diff
@@ -61,6 +61,7 @@ bool IIE::respond(const Bottle& command, Bottle& reply) {
         "quit \n";
 
     reply.clear();
+    reply.addVocab(Vocab::encode("many"));
```
Comment (Tobias-Fischer) [inline main/src/modules/tools/InsermInteractionEngine/src/iie.cpp]: As above, not sure what this is good for.

File: main/src/modules/abm/abmReasoning/src/abmReasoning.cpp (hunk 1 above)
Comment (gregoire-pointeau) [inline main/src/modules/abm/abmReasoning/src/abmReasoning.cpp]: It is used to display the help message not in one big line. Accessory and mainly for debug use. Can be removed.

File: main/src/modules/attentionRelated/proactiveTagging/src/proactiveTagging.cpp (hunk 3 above)
Comment (gregoire-pointeau) [inline main/src/modules/attentionRelated/proactiveTagging/src/proactiveTagging.cpp]: same

File: main/src/modules/abm/autobiographicalMemory/src/autobiographicalMemory.cpp (hunk 2 above)
Comment (gregoire-pointeau) [inline main/src/modules/abm/autobiographicalMemory/src/autobiographicalMemory.cpp]: Same

File: main/src/modules/reservoir/reservoirHandler/src/main.cpp (hunk 4 above)
Comment (gregoire-pointeau) [inline main/src/modules/reservoir/reservoirHandler/src/main.cpp]: Should I change it or did you did it?

File: main/src/modules/reservoir/reservoirHandler/include/reservoirHandler.h (hunk 5 above)
Comment (gregoire-pointeau) [inline main/src/modules/reservoir/reservoirHandler/include/reservoirHandler.h]: same.

File: main/src/modules/reservoir/reservoirHandler/src/reservoirHandler.cpp (hunk 6 above)
Comment (gregoire-pointeau) [inline main/src/modules/reservoir/reservoirHandler/src/reservoirHandler.cpp]: same

File: main/src/modules/tools/InsermInteractionEngine/src/iie.cpp (hunk 7 above)
Comment (gregoire-pointeau) [inline main/src/modules/tools/InsermInteractionEngine/src/iie.cpp]: same

File: main/src/modules/abm/abmReasoning/src/abmReasoning.cpp (hunk 1 above)
Comment (gregoire-pointeau) [inline main/src/modules/abm/abmReasoning/src/abmReasoning.cpp]: I tried it on some modules, and it didn't seem to mess anything (at least on `autobiographicalMemory` ). But I'm not 100% sure.

File: main/src/modules/abm/autobiographicalMemory/src/autobiographicalMemory.cpp (hunk 2 above)
Comment (gregoire-pointeau) [inline main/src/modules/abm/autobiographicalMemory/src/autobiographicalMemory.cpp]: Ok. I change, commit, push.

File: main/src/modules/reservoir/reservoirHandler/src/main.cpp (hunk 4 above)
Comment (gregoire-pointeau) [inline main/src/modules/reservoir/reservoirHandler/src/main.cpp]: done

File: main/src/modules/reservoir/reservoirHandler/include/reservoirHandler.h (hunk 5 above)
Comment (gregoire-pointeau) [inline main/src/modules/reservoir/reservoirHandler/include/reservoirHandler.h]: done

File: main/src/modules/reservoir/reservoirHandler/src/reservoirHandler.cpp (hunk 6 above)
Comment (gregoire-pointeau) [inline main/src/modules/reservoir/reservoirHandler/src/reservoirHandler.cpp]: done

**Comment by Tobias-Fischer** [comment]
Okay let's wait for @pattacini regarding the "many" thingy, and then I'm happy to merge :)

**Comment by pattacini** [comment]
As already said, `[main]` is used only for displaying messages nicely. It shouldn't be used in a module-module communication then.

**Comment by Tobias-Fischer** [comment]
Okay, please remove the `[main]` then @gregoire-pointeau, and I'll merge as soon as that's done.

**Comment by pattacini** [comment]
It seems that since the time **C9** has been acquired by Amazon, your credit card info are requested during subscription process - even if it's free.

Therefore, I've prepared another IDE in [**Codenvy**](https://codenvy.io), which is more user friendly from the privacy perspective 😉 .

If you sign up there I can share the IDE with you. Drop me an email once done.

cc @robotology/wysiwyd-developers

**Comment by gregoire-pointeau** [comment]
Ok, you can try now.

**Comment by Tobias-Fischer** [comment]
Hi @gregoire-pointeau,
It seems like this branch is still not rebased. Did you and @pattacini rebase the branch last week? Maybe you did a `git push --force` from another machine overwriting the rebase from last week?
Will merge as soon as it's rebased.

Best,
Tobi

**Comment by pattacini** [comment]
Another rebase is required because of your fix to `opcPopulator`, @Tobias-Fischer  😉

**Comment by gregoire-pointeau** [comment]
Ok, I didn't saw this one.

**Comment by Tobias-Fischer** [comment]
Uuuups, sorry for that, I completely forgot. In this case, I can do the rebase if you want. Let me know ;)

**Comment by pattacini** [comment]
Can you thus resolve the conflict and then finally merge the PR asap?
I don't want to deal with this branch any longer 😄

**Comment by gregoire-pointeau** [comment]
If you want to do it, I'll be glad ! help yourself @Tobias-Fischer !

**Comment by gregoire-pointeau** [comment]
"My name is Tobias Fisher, and I approve this pull request !"
Thanks to you two.

**Comment by Tobias-Fischer** [comment]
FisCher ;)
Just waiting for Travis to finish the build, will merge soon.

**Comment by pattacini** [comment]
Finally, we're done with this enormous PR 🎉 
It's been quite tough mainly because of the huge branch we had to merge.

Let's keep `feature` and `fix` branches more agile next time by breaking down big developments in multiple branches.

Further, it's also a good practice to rebase them often on top of `devel`.

**Pull request merged**